### PR TITLE
Np 47131 prevent persistence of invalid role names

### PR DESCRIPTION
**Pull request by truhacevkir**
We had a case where "old" role has still been attached to user in NVA. It caused that user had access rights we did not know about, cause role and all belongings access rights have not been removed/ updated when we updated user roles via api. 

To fix this case and all future cases we decided to make RoleName to enum, so all unknown values will fail on serialization. 

- Since roleName has been string before, had to update a lot of tests. 
- RoleName is still stored as String in database and all enum values have the same role names as before.

### CODE REVIEW

File: user-access-handlers/src/main/java/no/unit/nva/handlers/IdentityServiceMigrateCuratorHandler.java [hunk 1]
```diff
@@ -65,7 +65,7 @@ private UserDto updateRolesForUser(UserDto user) throws NotFoundException {
 
     private Set<RoleDto> updateRoleSet(Set<RoleDto> roleSet) {
         roleSet.add(NVI_CURATOR_ROLE);
-        roleSet.removeIf(role -> LEGACY_NVI_CURATOR_NAME.equals(role.getRoleName()));
+        roleSet.removeIf(role -> DEPRECATED_NVI_CURATOR.equals(role.getRoleName()));
```
Comment (truhacevkir) [inline user-access-handlers/src/main/java/no/unit/nva/handlers/IdentityServiceMigrateCuratorHandler.java:68]: Had to add "old" nvi-curator role to be able to run migration test

File: user-access-internal-model/src/test/java/no/unit/nva/useraccessservice/dao/RoleDbTest.java [hunk 2]
```diff
@@ -112,38 +110,38 @@ void buildWithoutRoleNameShouldThrowException() {
...
-        assertThat(exception.getMessage(), containsString(EMPTY_ROLE_NAME_ERROR));
-    }
+//    @ParameterizedTest(name = "setPrimaryHashKey throws exception when input is:\"{0}\"")
+//    @NullAndEmptySource
+//    @ValueSource(strings = {" ", "\t", "\n", "\r"})
+//    void setPrimaryHashKeyThrowsExceptionWhenInputIsBlankOrNullString(String blankString) {
+//        Executable action = () -> RoleDb.newBuilder().withName(blankString).build();
+//        InvalidEntryInternalException exception = assertThrows(InvalidEntryInternalException.class, action);
+//        assertThat(exception.getMessage(), containsString(EMPTY_ROLE_NAME_ERROR));
+//    }
```
Comment (truhacevkir) [inline user-access-internal-model/src/test/java/no/unit/nva/useraccessservice/dao/RoleDbTest.java:132]: Following test is not longer relevant

File: user-access-internal-model/src/test/java/no/unit/nva/useraccessservice/dao/UserDaoTest.java [hunk 3]
```diff
@@ -174,42 +175,42 @@ void shouldConvertToDtoAndBackWithoutInformationLoss() {
...
+//        throws InvalidEntryInternalException {
+//        RoleDto invalidRole = RoleDto.newBuilder().withRoleName(SOME_ROLENAME).build();
+//        invalidRole.setRoleName(invalidRoleName);
+//        List<RoleDto> invalidRoles = Collections.singletonList(invalidRole);
+//        UserDto userWithInvalidRole = UserDto.newBuilder().withUsername(SOME_USERNAME).withRoles(invalidRoles).build();
+//
+//        Executable action = () -> UserDao.fromUserDto(userWithInvalidRole);
+//        RuntimeException exception = assertThrows(RuntimeException.class, action);
+//        assertThat(exception.getCause(), is(instanceOf(InvalidEntryInternalException.class)));
+//    }
```
Comment (truhacevkir) [inline user-access-internal-model/src/test/java/no/unit/nva/useraccessservice/dao/UserDaoTest.java:207]: Unsure if we still need these tests, all the unknown enum values will from now on be illegal.

File: user-access-public-model/src/test/java/no/unit/nva/useraccessservice/model/RoleDtoTest.java [hunk 4]
```diff
@@ -50,40 +48,33 @@ void roleDtoShouldHaveABuilder() {
...
     }
 
-    @ParameterizedTest(name = "builder should throw exception when rolename is:\"{0}\"")
-    @NullAndEmptySource
-    @ValueSource(strings = {"", " "})
-    void builderShouldNotAllowEmptyRoleName(String rolename) {
-        Executable action = () -> RoleDto.newBuilder().withRoleName(rolename).build();
-        assertThrows(InvalidEntryInternalException.class, action);
-    }
-
```
Comment (truhacevkir) [inline user-access-public-model/src/test/java/no/unit/nva/useraccessservice/model/RoleDtoTest.java:74]: Not relevant test longer

**Comment by truhacevkir** [comment]
From now we will end up with:

java.lang.IllegalArgumentException: 
Unable to convert string value 'Curator' to enum type 'class no.unit.nva.useraccessservice.model.RoleName'

And return internal server error via api.

Before:
Not longer "existent" role has been served via api and user got granted access rights to given role:
```
{
    "rolename" : "Curator",
    "accessRights" : [ "MANAGE_PUBLISHING_REQUESTS", "MANAGE_OWN_AFFILIATION", "MANAGE_DOI", "MANAGE_RESOURCES_ALL", "MANAGE_RESOURCES_STANDARD" ],
    "type" : "Role"
  }
```

**Review by sondrev — APPROVED** [review]
Looks good

**Review by anetteOlli — COMMENTED** [review]
Too long, didn't read.

Quick question what will happen to a user that has one of the legacy userRoles when they log in now? Do we need to wash our databases before we roll this to prod?

**Comment by truhacevkir** [comment]
> Too long, didn't read. (I am sure it just a lot of refactor from a string to enum so its fine).
> 
> Quick question what will happen to a user that has one of the legacy userRoles when they log in now? Do we need to wash our databases before we roll this to prod?

All "old" roles will be removed from users.

**Review by anetteOlli — DISMISSED** [review]

**Pull request merged**